mesh/mesh.go

```diff
@@ -239,23 +232,66 @@ func (msh *Mesh) GetLayer(i types.LayerID) (*types.Layer, error) {
 	return l, nil
 }
 
+type internalLayer struct {
+	layerID   types.LayerID
+	layerHash types.Hash32
+}
+
 type validator struct {
 	*Mesh
-	processedLayer types.LayerID
+	processedLayer internalLayer
 }
 
 func (vl *validator) ProcessedLayer() types.LayerID {
-	defer vl.lvMutex.RUnlock()
-	vl.lvMutex.RLock()
+	return vl.getProcessedLayer().layerID
+}
+
+func (vl *validator) ProcessedLayerHash() types.Hash32 {
+	return vl.getProcessedLayer().layerHash
+}
+
+func (vl *validator) SetProcessedLayer(layerID types.LayerID, hash types.Hash32) {
+	layer := &internalLayer{
+		layerID:   layerID,
+		layerHash: hash,
+	}
+	vl.setProcessedLayer(layer)
+}
+
+func (vl *validator) getProcessedLayer() internalLayer {
+	vl.mutex.RLock()
+	defer vl.mutex.RUnlock()
 	return vl.processedLayer
 }
 
-func (vl *validator) SetProcessedLayer(lyr types.LayerID) {
-	vl.With().Info("set processed layer", lyr)
+func (vl *validator) setProcessedLayer(lyr *internalLayer) {
+	vl.mutex.Lock()
+	defer vl.mutex.Unlock()
+	if vl.processedLayer.layerID.Add(1) != lyr.layerID {
+		vl.With().Error("could not set processed layer",
+			log.FieldNamed("old_processed_layer", vl.processedLayer.layerID),
+			log.FieldNamed("new_processed_layer", lyr.layerID))
+		return
+	}
+	vl.processedLayer = *lyr
 	events.ReportNodeStatusUpdate()
-	defer vl.lvMutex.Unlock()
-	vl.lvMutex.Lock()
-	vl.processedLayer = lyr
+	vl.Event().Info("processed layer set", lyr.layerID, log.String("layer_hash", lyr.layerHash.Hex()))
+}
+
+func (vl *validator) persistProcessedLayer() {
+	lyr := vl.getProcessedLayer()
+	data, err := types.InterfaceToBytes(lyr)
```

Comment:
`InterfaceToBytes` only serialises exported fields, in our case  `internalLayer`s fields are unexported 
Please add a unit test to verify persist and store

Reply:
done